### DISCUSSION TITLE
How to customise the open telemetry for logs Provider?

### DISCUSSION THREAD

**Discussion by p10tyr** [open-telemetry/opentelemetry-dotnet#3275] [Q&A] [answered]
I am referring to this page  https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/logs/customizing-the-sdk that has TODO around the customisation of the log providers.

I have several of my own LogProviders made already that ALWAYS inject specific things about the application so that we can group, query and analyse by our business needs. The implementation is always specific to the product we integrating.

> I am generally interested in how to add my own Attributes to the LogRecord.

----
I can achieve the same using custom or extending the LogProvider but I am still stuck with the LogRecord that I can't really add too much to add.

But its unclear how to customise the OpenTelemetryLoggerProvider

----
So I have been looking at adding a custom log processor but its unclear how I can use DI with this.

Ideally I would like to add extra attributes to **LogRecord** but there is northing to allow me to do this?

The[ OpenTelemetry data model](https://opentelemetry.io/docs/reference/specification/logs/data-model/) for stand alone logs suggest that there _should an Attributes property_  I am pretty sure this is where the current States KV are being copied over to.. but how does a 1st class application add their own attributes?

```
 openTelemetryLoggerOptions.AddProcessor(new MyExporter());
...
 internal class MyExporter : BaseProcessor<LogRecord>
 {
     public override void OnStart(LogRecord logRecord) {}
     public override void OnEnd(LogRecord logRecord) { **? How do I add my own Attributes here**? }
 }
```

---
Ideally I would like to add extra attributes to LogRecord but I can not see any options to do so from the above methods.

So the only other place I can see how to add attributes is to dynamically set it Resource builder?
But I understand that this is static once of initialisation, like ServiceName, Machine Name.. stuff that WONT change until restarted.(unless I am mistaken?)

This is ideal for me because I can add my attribute to everything being exported - I just need the value to be dynamic per request.

```
var otelAttributes = new List<KeyValuePair<string, object>>();
otelAttributes.Add(new KeyValuePair<string, object>("deployment.environment", ResourceNameHelper.GetEnvironmentName()));
otelAttributes.Add(new KeyValuePair<string, object>(Dynamic, Value); <-- ??

return ResourceBuilder
       .CreateDefault()
       .AddService(ResourceNameHelper.ServiceName)
      .AddAttributes(otelAttributes);
```

**Comment by cijothomas** [comment, marked as answer]
Is your goal to add more attributes to logs, but without modifying the ilogger.log().. call itself?
Have you tried using Scopes?

**Reply by CodeBlanch** [reply]
@ppumkin Just to expand on that a bit...

Here is the code inside of OtlpExporter:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/e7f06afd51489e3047f167d9b7362aba512508c0/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs#L144-L156

It will loop through scopes and automatically add to the attributes.

There is an option for controlling whether or not scopes are included: https://github.com/open-telemetry/opentelemetry-dotnet/blob/e7f06afd51489e3047f167d9b7362aba512508c0/src/OpenTelemetry/Logs/OpenTelemetryLoggerOptions.cs#L33

If you turn that on, you should be able to push the tenant information via middleware like this:

```csharp
    internal sealed class TenantLogEnrichmentMiddleware
    {
        private readonly ILogger<TenantLogEnrichmentMiddleware> _Logger;
        private readonly ITenantResolver _TenantResolver;
        private readonly RequestDelegate _Next;

        public TenantLogEnrichmentMiddleware(
            ILogger<TenantLogEnrichmentMiddleware> logger,
            ITenantResolver tenantResolver,
            RequestDelegate next)
        {
            _Logger = logger ?? throw new ArgumentNullException(nameof(logger));
            _TenantResolver = tenantResolver ?? throw new ArgumentNullException(nameof(tenantResolver));
            _Next = next ?? throw new ArgumentNullException(nameof(next));
        }

        public async Task InvokeAsync(HttpContext httpContext)
        {
            var tenant = _TenantResolver.Tenant; // Some kind of scoped service which understands how to resolve the tenant for the current request
            using var scope = _Logger.BeginScope(tenant);
            await _Next(httpContext).ConfigureAwait(false);
        }
    }
```

Just make sure whatever you pass to `BeginScope` implements `IReadOnlyList<KeyValuePair<string, object>>` with the values you want to push into attributes.

**Reply by p10tyr** [reply]
I think my main issue was tunnel vision and I only say my own code and my own solutions. Thank you for providing the middleware code because this made me realise my mistake. I thought you only had to begin scope in the services we wrote logs which for us is like 20k lines of log code. 

So unfortunately I amusing Framework still but the concept is the same. I just start the scope on the HttpRequest which solves all the Dependency Injection problems and NOW it makes sense why Scopes are being copied to attributes!

Thank you very much guys!

**Reply by p10tyr** [reply]
Although just realised that version rc9.2 does not support copying Scopes in the Logger package??

I have enabled 
```
options.IncludeScopes = true;
options.IncludeFormattedMessage = true;
options.ParseStateValues = true;
```
And created an ActionFilter attribute that I just attached to everything in Global. That does basically what you mentioned in the middle ware but in the Filter

```
var attributeScopes = new List<KeyValuePair<string, object>>();
attributeScopes.Add(new KeyValuePair<string, object>("ehs_clientid", tenant.ClientId));

using var scope = logger.BeginScope(attributeScopes);

base.OnActionExecuting(actionContext);
```

I have verified that the Scope has this for every single request because our existing log provider shows it attached in the Scope. I have also debugged the provider and it works as expected in terms of keeping the Scope on the request. 

But in Elastic Search using Open Telemetry Exporter using the same test error I can NOT see the `labels.ehs_clientid` being attached but I do see the states of the error message formatter. 

![image](https://user-images.githubusercontent.com/1249683/168286874-6f9b655f-4dec-4871-b17a-f9edccc22f08.png)

**Reply by CodeBlanch** [reply]
Woops sorry @ppumkin it looks like scope support was added in [OtlpExporter very recently](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3218). It should be available in the next beta soon. Not sure exactly when @cijothomas is thinking to release that, but probably within the next week or two.
